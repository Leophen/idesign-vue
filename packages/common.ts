import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
