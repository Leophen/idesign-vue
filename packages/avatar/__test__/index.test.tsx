import { Avatar } from '../index';
import { render } from '../../common'

const imageSrc = 'https://picsum.photos/180/120';

describe('Avatar 组件测试', () => {
  it('image src', () => {
    const wrapper = render(<Avatar image={imageSrc} />);
    const image = wrapper.find('.i-avatar__image');
    expect(image.attributes('src')).toBe(imageSrc);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('children', () => {
    const wrapper = render(<Avatar>L</Avatar>);
    expect(wrapper.find('.i-avatar').text()).toBe('L');
  });

  it('size', () => {
    const wrapper = render(<Avatar size={24}>L</Avatar>);
    expect(wrapper.element.getAttribute('style')).toContain('width: 24px');
  });

  it('shape', () => {
    const wrapper = render(<Avatar shape="round">L</Avatar>);
    expect(wrapper.classes()).toContain('i-avatar__shape-round');
  })

  it('color', () => {
    const wrapper = render(<Avatar color="rgb(50, 54, 164)">L</Avatar>);
    expect(wrapper.element.getAttribute('style')).toMatch('background: rgb(50, 54, 164);');
  });
});



