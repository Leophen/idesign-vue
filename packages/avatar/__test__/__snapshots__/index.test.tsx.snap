// Vitest Snapshot v1

exports[`Avatar 组件测试 > image alt 1`] = `
<div
  class="i-avatar"
  style="width: 32px; height: 32px;"
>
  <img
    class="i-avatar__image"
    referrerpolicy="no-referrer"
    src="https://picsum.photos/180/120"
  />
  <!--v-if-->
  <!--v-if-->
</div>
`;

exports[`Avatar 组件测试 > image src 1`] = `
<div
  class="i-avatar"
  style="width: 32px; height: 32px;"
>
  <img
    class="i-avatar__image"
    referrerpolicy="no-referrer"
    src="https://picsum.photos/180/120"
  />
  <!--v-if-->
  <!--v-if-->
</div>
`;
